crates/sui-core/src/authority/execution_time_estimator.rs

```diff
@@ -0,0 +1,777 @@
+                observer
+                    .record_local_observations(&tx, &timings, total_duration)
+                    .await;
+            }
+            info!("shutting down ExecutionTimeObserver");
+        });
+    }
+
+    #[cfg(test)]
+    fn new_for_testing(
+        epoch_store: Arc<AuthorityPerEpochStore>,
+        consensus_adapter: Box<dyn SubmitToConsensus>,
+    ) -> Self {
+        Self {
+            epoch_store: Arc::downgrade(&epoch_store),
+            consensus_adapter,
+            local_observations: HashMap::new(),
+        }
+    }
+
+    // Used by execution to report observed per-entry-point execution times to the estimator.
+    // Updates moving averages and submits observation to consensus if local observation differs
+    // from consensus median.
+    // TODO: Consider more detailed heuristic to account for overhead outside of commands.
+    async fn record_local_observations(
+        &mut self,
+        tx: &ProgrammableTransaction,
+        timings: &[ExecutionTiming],
+        total_duration: Duration,
+    ) {
+        assert_eq!(tx.commands.len(), timings.len());
+
+        let total_command_duration: Duration = timings.iter().map(|t| t.duration()).sum();
+        let extra_overhead = total_duration - total_command_duration;
+
+        let mut to_share = Vec::with_capacity(tx.commands.len());
+        for (i, command) in tx.commands.iter().enumerate() {
+            // TODO: Consider using failure/success information in computing estimates.
+            let mut command_duration = timings[i].duration();
+
+            // Distribute overhead proportionally to each command's measured duration.
+            command_duration = command_duration
+                + extra_overhead
+                    .mul_f64(command_duration.as_secs_f64() / total_command_duration.as_secs_f64());
+
+            // For native commands, adjust duration by length of command's inputs/outputs.
+            // This is sort of arbitrary, but hopefully works okay as a heuristic.
+            command_duration = command_duration.div_f64(command_length(command));
+
+            // Update moving-average observation for the command.
+            let key = ExecutionTimeObservationKey::from_command(command);
+            let local_observation =
+                self.local_observations
+                    .entry(key.clone())
+                    .or_insert_with(|| LocalObservations {
+                        moving_average: SingleSumSMA::from_zero(Duration::ZERO),
+                        last_shared: None,
+                    });
+            local_observation
+                .moving_average
+                .add_sample(command_duration);
+
+            // Send a new observation through consensus if our current moving average
+            // differs too much from the last one we shared.
+            // TODO: Consider only sharing observations for entrypoints with congestion.
```

Comment:
there is a further TODO to note, which is that we shouldn't bother sharing if the consensus estimate already agrees with our local estimate

Reply:
that would require adding a dependency on the estimator to the observer, which is something I was trying to avoid, but yes it's something we def could consider.
updated the todo